File: Leetcode/MinimumTimeVisitingAllPoints.rs
// https://leetcode.com/problems/minimum-time-visiting-all-points

struct Solution {}

impl Solution {
    pub fn min_time_to_visit_all_points(points: Vec<Vec<i32>>) -> i32 {
        points
            .windows(2)
            .map(|pair| {
                let x1 = pair[0][0];
                let x2 = pair[1][0];
                let y1 = pair[0][1];
                let y2 = pair[1][1];

                let dx = (x1 - x2).abs();
                let dy = (y1 - y2).abs();

                dx + dy - dx.min(dy)
            })
            .sum()
    }
}

File: Leetcode/MinimumTimeVisitingAllPoints.cpp
// https://leetcode.com/problems/minimum-time-visiting-all-points

class Solution {
   public:
    int minTimeToVisitAllPoints(vector<vector<int>>& points) {
        int ans = 0;
        int x = points[0][0];
        int y = points[0][1];

        for (int i = 1; i < points.size(); ++i) {
            int xto = points[i][0];
            int yto = points[i][1];

            ans += max(abs(x - xto), abs(y - yto));
            x = xto, y = yto;
        }

        return ans;
    }
};